tests/global_communication/test_communicator.cpp

```diff
@@ -42,97 +50,120 @@ struct source_proxy {
-        const auto n_local_cells = n_local_spikes;
+    const auto n_local_spikes = 10;
+    const auto n_local_cells = n_local_spikes;
 
-        // Important: set up meta-data in dry run back end.
+    // Important: set up meta-data in dry run back end.
+    if (is_dry_run()) {
         policy::set_sizes(policy::size(), n_local_cells);
+    }
 
-        // create local spikes for communication
-        std::vector<spike_proxy> local_spikes;
-        for (auto i=0; i<n_local_spikes; ++i) {
-            local_spikes.push_back(spike_proxy{i, rank});
-        }
+    // Create local spikes for communication
+    std::vector<spike_proxy> local_spikes;
+    for (auto i=0; i<n_local_spikes; ++i) {
+        local_spikes.push_back(spike_proxy{i, rank});
+    }
 
-        // perform exchange
-        const auto global_spikes = policy::gather_spikes(local_spikes);
+    // Perform exchange
+    const auto global_spikes = policy::gather_spikes(local_spikes);
 
-        // test that partition information is correct
-        const auto& part = global_spikes.partition();
-        EXPECT_EQ(num_domains+1u, part.size());
-        for (auto i=0u; i<part.size(); ++i) {
-            EXPECT_EQ(part[i], n_local_spikes*i);
-        }
+    // Test that partition information is correct
+    const auto& part = global_spikes.partition();
+    EXPECT_EQ(num_domains+1u, part.size());
+    for (auto i=0u; i<part.size(); ++i) {
+        EXPECT_EQ(part[i], n_local_spikes*i);
+    }
 
-        // test that spikes were correctly exchanged
-        //
-        // The local spikes had sources numbered 0:n_local_spikes-1
-        // The global exchange should replicate the local spikes and
-        // shift their sources to make them local to the "dummy" source
-        // domain.
-        // We set the model up with n_local_cells==n_local_spikes with
-        // one spike per local cell, so the result of the global exchange
-        // is a list of num_domains*n_local_spikes spikes that have
-        // contiguous source gid
-        const auto& spikes = global_spikes.values();
-        for (auto i=0u; i<spikes.size(); ++i) {
-            const auto s = spikes[i];
-            EXPECT_EQ(i, unsigned(s.source.gid));
-            EXPECT_EQ(0, s.domain);
-        }
+    // Test that spikes were correctly exchanged
+    //
+    // The local spikes had sources numbered 0:n_local_spikes-1
+    // The global exchange should replicate the local spikes and
+    // shift their sources to make them local to the "dummy" source
+    // domain.
+    // We set the model up with n_local_cells==n_local_spikes with
+    // one spike per local cell, so the result of the global exchange
+    // is a list of num_domains*n_local_spikes spikes that have
+    // contiguous source gid
+    const auto& spikes = global_spikes.values();
+    for (auto i=0u; i<spikes.size(); ++i) {
+        const auto s = spikes[i];
+        EXPECT_EQ(i, unsigned(s.source.gid));
+        EXPECT_EQ(0, s.domain);
```

Comment:
I have fixed it. It required 2 changes
  1. mpi mode has to explicitly set the source gid to the correct gid (instead of relying on the communicator back end creating the remote gid correctly)
  2. the check must do the test you suggest when in mpi mode.